src/ngx_http_lua_worker_thread.c

```diff
@@ -0,0 +1,497 @@
+    lua_State                          *vm;
+    size_t                              len;
+    const char                         *mod_name;
+    const char                         *func_name;
+    int                                 rc;
+    const char                         *err;
+    int                                 i;
+    ngx_thread_task_t                  *task;
+    ngx_http_lua_worker_thread_ctx_t   *worker_thread_ctx;
+
+    r = ngx_http_lua_get_req(L);
+
+    if (r == NULL) {
+        return luaL_error(L, "no request found");
+    }
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+
+    if (ctx == NULL) {
+        return luaL_error(L, "no ctx found");
+    }
+
+    ngx_http_lua_check_context(L, ctx, NGX_HTTP_LUA_CONTEXT_YIELDABLE);
+
+    n_args = lua_gettop(L);
+
+    if (n_args < 3) {
+        lua_pushboolean(L, 0);
+        lua_pushstring(L, "expecting at least 3 arguments");
+        return 2;
+    }
+
+    thread_pool_name.data = (u_char*) lua_tolstring(L, 1,
+                                                    &thread_pool_name.len);
+    thread_pool = ngx_thread_pool_get((ngx_cycle_t *) ngx_cycle,
+                                      &thread_pool_name);
+    if (thread_pool == NULL) {
+        lua_pushboolean(L, 0);
+        lua_pushstring(L, "thread_pool not found");
+        return 2;
+    }
+
+    /* get vm */
+    tctx = ngx_http_lua_get_task_ctx(L, r);
+    if (tctx == NULL) {
+        lua_pushboolean(L, 0);
+        lua_pushstring(L, "no available Lua vm");
+        return 2;
+    }
+    vm = tctx->vm;
+
+    /* push function from module require */
+    mod_name = lua_tolstring(L, 2, &len);
+    func_name = lua_tolstring(L, 3, NULL);
+    lua_getfield(vm, LUA_GLOBALSINDEX, "require");
+    lua_pushlstring(vm, mod_name, len);
+    rc = lua_pcall(vm, 1, 1, 0);
+
+    if (rc != 0) {
+        err = lua_tolstring(vm, 1, &len);
+        lua_pushboolean(L, 0);
+        lua_pushlstring(L, err, len);
+        ngx_http_lua_free_task_ctx(tctx);
+        return 2;
+    }
+
+    if (!lua_istable(vm, -1)) {
+        lua_pushboolean(L, 0);
+        lua_pushstring(L, "invalid lua module");
+        ngx_http_lua_free_task_ctx(tctx);
+        return 2;
+    }
+
+    lua_getfield(vm, -1, func_name);
+    if (!lua_isfunction(vm, -1)) {
+        lua_pushboolean(L, 0);
+        lua_pushstring(L, "invalid function");
+        ngx_http_lua_free_task_ctx(tctx);
+        return 2;
+    }
+
+    /* remove the table returned by require */
+    lua_remove(vm, 1);
+
+    /* copying passed arguments */
+    for (i = 4; i <= n_args; i++) {
+        if (ngx_http_lua_xcopy(L, vm, i, 1) == LUA_TNONE) {
+            lua_pushboolean(L, 0);
+            lua_pushstring(L, "unsupported argument type");
+            ngx_http_lua_free_task_ctx(tctx);
+            return 2;
+        }
+    }
+
+    /* post task */
+    task = ngx_http_lua_thread_task_alloc(
+        sizeof(ngx_http_lua_worker_thread_ctx_t));
```

Comment:
```suggestion
                sizeof(ngx_http_lua_worker_thread_ctx_t));
```
this style can be better.